### PR TITLE
Implicit IB milestone 1: Restore implicit IB solver to a functioning state

### PR DESCRIPTION
**Pull request by boyceg**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format (i.e., `make indent`) run? For more information see
      `scripts/formatting/README.md`.
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by boyceg** [comment]
This is part one of a series of PRs to get @colegruninger97's coupling-aware Vanka preconditioner into IBAMR. I have a roadmap of what I think we should try to do in doc.

File: tests/IB/implicit_stokes_ib_01.baseline.output [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+IBHierarchyIntegrator::integrateHierarchy_velocity(): interpolating Eulerian velocity to the Lagrangian mesh
+IBHierarchyIntegrator::postprocessIntegrateHierarchy(): interpolating Eulerian velocity to the Lagrangian mesh
+IBHierarchyIntegrator::postprocessIntegrateHierarchy(): CFL number = 0
+IBHierarchyIntegrator::postprocessIntegrateHierarchy(): Eulerian estimate of upper bound on IB point displacement since last regrid = 0
+IBHierarchyIntegrator::advanceHierarchy(): synchronizing updated data
+IBHierarchyIntegrator::advanceHierarchy(): resetting time dependent data
+
+At end       of timestep # 1
+Simulation time is 0.01
++++++++++++++++++++++++++++++++++++++++++++++++++++
```
Comment (drwells) [inline tests/IB/implicit_stokes_ib_01.baseline.output:244]: What's the intent with this test? Since it only does one timestep everything is zero.

File: plans/implicit-ib/roadmap.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+Focus on a staged roadmap to stabilize implicit IB, validate algorithmic correspondence, improve performance, and expand capability (parallel + kernel generalization).
+Use deterministic CMake tests as the main confidence signal, but treat milestone checks as directional goals rather than strict gates.
+
+## Roadmap Record
+- Store and maintain roadmap at: `/Users/boyceg/code/IBAMR/doc/implicit-ib-roadmap.md`
+- Keep this file updated with:
+1. Current baseline configs
+2. Milestones and status
+3. Key comparison notes vs MATLAB
+4. PR history and follow-up items
```
Comment (drwells) [inline plans/implicit-ib/roadmap.md:13]: This is a useful record of what we are doing but (like the other PR) it isn't something which would go in the manual, so I think we should put it somewhere else.

**Review by drwells — CHANGES_REQUESTED** [review]
Nice!

File: tests/IB/implicit_stokes_ib_01.baseline.output (hunk 1 above)
Comment (boyceg) [reply tests/IB/implicit_stokes_ib_01.baseline.output:244]: It runs the implicit solver for one time step to make sure it doesn't crash.

File: plans/implicit-ib/roadmap.md (hunk 2 above)
Comment (boyceg) [reply plans/implicit-ib/roadmap.md:13]: Sure.

Comment (boyceg) [reply plans/implicit-ib/roadmap.md:13]: Where do you want this kind of stuff?

Comment (drwells) [reply plans/implicit-ib/roadmap.md:13]: Some other directory - perhaps `/ai/` or `/plans/`. Given codex's fascination with the word baselines, we could do `/baselines/` too :)

Comment (boyceg) [reply plans/implicit-ib/roadmap.md:13]: Sounds good.

I can tell it to knock it off with this baselines nonsense. 😀

**Comment by boyceg** [comment]
It looks like an `IBInstrumentPanel` is being created somewhere in the implicit solver. Let me see if I can get rid of that...

File: include/ibamr/INSStaggeredHierarchyIntegrator.h [hunk 3]
```diff
@@ -396,6 +396,7 @@ class INSStaggeredHierarchyIntegrator : public INSHierarchyIntegrator
      */
     SAMRAI::tbox::Pointer<StaggeredStokesPhysicalBoundaryHelper> d_bc_helper;
     SAMRAI::tbox::Pointer<IBTK::SideDataSynchronization> d_U_rhs_side_synch_op;
+    int d_U_rhs_synch_data_idx = IBTK::invalid_index;
```
Comment (drwells) [inline include/ibamr/INSStaggeredHierarchyIntegrator.h]: Could you add some documentation for this value?

File: src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp [hunk 4]
```diff
@@ -1567,9 +1567,18 @@ INSStaggeredHierarchyIntegrator::setupSolverVectors(const Pointer<SAMRAIVectorRe
...
+    if (rhs_synch_data_idx != d_U_rhs_synch_data_idx)
+    {
+        d_U_rhs_side_synch_op->deallocateOperatorState();
+        d_U_rhs_side_synch_op->initializeOperatorState(rhs_synch_transaction, d_hierarchy);
+        d_U_rhs_synch_data_idx = rhs_synch_data_idx;
+    }
+    else
+    {
+        d_U_rhs_side_synch_op->resetTransactionComponent(rhs_synch_transaction);
+    }
```
Comment (drwells) [inline src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp]: I missed this the first time. Is this going to be run for everything, or just implicit IB? Would it be possible to move this over to the implicit IB integrator?

**Review by drwells — CHANGES_REQUESTED** [review]
Two more comments. Could you make it clear why (and if) we have to modify the INS solver so this implicit IB solver works?

File: include/ibamr/INSStaggeredHierarchyIntegrator.h (hunk 3 above)
Comment (boyceg) [reply include/ibamr/INSStaggeredHierarchyIntegrator.h]: I was able to remove this.

File: src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp (hunk 4 above)
Comment (boyceg) [reply src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp]: I was able to remove this.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp [hunk 5]
```diff
@@ -794,8 +781,6 @@ IBImplicitStaggeredHierarchyIntegrator::integrateHierarchy_velocity(const double
     ierr = MatDestroy(&interp_op);
     IBTK_CHKERRQ(ierr);
 
-    // Execute any registered callbacks.
-    executeIntegrateHierarchyCallbackFcns(current_time, new_time, cycle_num);
```
Comment (drwells) [inline src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp:798]: I also missed this one before: where are the callbacks now?

File: scripts/formatting/indent_common.sh [hunk 6]
```diff
@@ -283,19 +283,22 @@ ensure_bracket_includes()
...
+    {
+      if ($0 ~ /^#[[:space:]]*include[[:space:]]*<[^>]+>$/) {
+        header = $0
+        sub(/^#[[:space:]]*include[[:space:]]*</, "", header)
+        sub(/>$/, "", header)
+        if (header in local_headers) sub(/<[^>]+>/, "\"" header "\"")
+      }
+      print
+    }
+  ' <(find "$(dirname "$1")" -type f -exec basename {} \;) "${tmpfile1}" > "${tmpfile2}"
```
Comment (drwells) [inline scripts/formatting/indent_common.sh]: Sorry about making a mess here - I forgot that the CI wasn't set up to actually run these scripts on macs.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp [hunk 7]
```diff
@@ -203,21 +215,51 @@ IBImplicitStaggeredHierarchyIntegrator::preprocessIntegrateHierarchy(const doubl
...
+        d_eul_sol_vec->addComponent(p_var, p_scratch_idx, wgt_cc_idx, d_hier_pressure_data_ops);
+
+        d_eul_rhs_vec = d_eul_sol_vec->cloneVector(d_object_name + "::eulerian_rhs_vec");
+        d_u_scratch_vec = d_eul_sol_vec->cloneVector(d_object_name + "::u_scratch_vec");
+        d_f_scratch_vec = d_eul_rhs_vec->cloneVector(d_object_name + "::f_scratch_vec");
+    }
+
+    d_eul_rhs_vec->allocateVectorData(current_time);
+    d_u_scratch_vec->allocateVectorData(current_time);
+    d_f_scratch_vec->allocateVectorData(current_time);
```
Comment (abarret) [inline src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp]: Suggest moving this to its own member function.

**Review by abarret — COMMENTED** [review]
It seems the only code change is to make the solver vec's member variables.

The roadmap seems reasonable. Is RT0 prolongation important for solver performance? Is the intent for the first few milestones to only test the application of the smoother, without prolongation or restrictions?

**Comment by boyceg** [comment]
> The roadmap seems reasonable. Is RT0 prolongation important for solver performance?

RT0 is important. @colegruninger97 can you send the references?

> Is the intent for the first few milestones to only test the application of the smoother, without prolongation or restrictions?

This is mostly getting @colegruninger97's solver from his PhD thesis implemented. The whole thing was broken, though — hence this commit.

All of the prolongation and restriction infrastructure is already in place, but was never carefully tested. Like after this first PR, the implicit solver "works" — it just isn't competitive with an explicit solver.

The goal of this project is to see if we can come up with an implicit solver that is competitive with an explicit one — hopefully faster.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp (hunk 7 above)
Comment (boyceg) [reply src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp]: Done.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp (hunk 5 above)
Comment (boyceg) [reply src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp:798]: Added back to the common integrateHierarchy() implementation.

At some point I am going to split the two modes (velocity vs. position-based), but not yet.

**Comment by boyceg** [comment]
FWIW, we are set up to try both RT0 and linear prolongation in IBAMR along with the corresponding adjoint restriction.

File: doc/news/changes/minor/20260301Griffith [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Stabilization: Restore baseline implicit Stokes-IB integration workflow and promote a finished attest test for the implicit hierarchy path.
```
Comment (drwells) [inline doc/news/changes/minor/20260301Griffith:1]: Can you make this more in line with the way we usually write changelog entries?

I'm not sure we should add a changelog at this point since it will be superseded by "implicit IB now works" later.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp [hunk 9]
```diff
@@ -794,11 +750,52 @@ IBImplicitStaggeredHierarchyIntegrator::integrateHierarchy_velocity(const double
...
-    executeIntegrateHierarchyCallbackFcns(current_time, new_time, cycle_num);
     return;
 } // integrateHierarchy_velocity
 
+void
+IBImplicitStaggeredHierarchyIntegrator::setupSolverVectors(const double current_time,
+                                                           const int coarsest_ln,
+                                                           const int finest_ln)
+{
+    Pointer<INSStaggeredHierarchyIntegrator> ins_hier_integrator = d_ins_hier_integrator;
```
Comment (drwells) [inline src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp:761]: Since this is also a 'staggered'  class, can we just store a pointer with the correct type?

File: tests/IB/implicit_stokes_ib_01.cpp [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+            iteration_num += 1;
+        }
+
+        // Cleanup Eulerian boundary condition specification objects (when
+        // necessary).
+        for (unsigned int d = 0; d < NDIM; ++d) delete u_bc_coefs[d];
+
+    } // cleanup dynamically allocated objects prior to shutdown
+
+    return 0;
```
Comment (drwells) [inline tests/IB/implicit_stokes_ib_01.cpp:279]: Do you anticipate that we will reuse this same test executable for future tests?

**Review by drwells — CHANGES_REQUESTED** [review]
I have a few minor comments.

File: doc/news/changes/minor/20260301Griffith (hunk 8 above)
Comment (boyceg) [reply doc/news/changes/minor/20260301Griffith:1]: Before this branch, the solver was crashing when I tried to use it, and so this is bringing things back to a baseline working (if still inefficient) state and adding a test to make sure that it doesn't crash.

File: src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp (hunk 9 above)
Comment (boyceg) [reply src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp:761]: I think that I would rather use the more general base class. I think that the position-based implicit method doesn't formally require a staggered solver, and so at some point I think I'd like to split them into two classes — one that needs this specialized staggered-grid multigrid algorithm, and another that (may) not.

File: tests/IB/implicit_stokes_ib_01.cpp (hunk 10 above)
Comment (boyceg) [reply tests/IB/implicit_stokes_ib_01.cpp:279]: Yes of course. Also, AFAIK, there is no test coverage for this class in `master` at present, so we need to do something.

**Comment by boyceg** [comment]
Milestone branch 5 (#1904) got pretty messy so I decided to do these (#1919 -- #1926).
